From e6bf58422310fb04ed49b60db572bba0c61ff1ae Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@foss.st.com>
Date: Mon, 21 Oct 2024 14:29:19 +0200
Subject: [PATCH] net: ethernet: stmmac: add grant/release of stm32 firewall

If applicable, ensure that stm32 firewall is properly configured
prior to accessing to DWMAC registers.

Change-Id: I519b9f87381c8ad098349573456c10878e995d73
Signed-off-by: Christophe Roullier <christophe.roullier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/412798
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 .../net/ethernet/stmicro/stmmac/dwmac-stm32.c | 30 +++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c
index 405399377d00..9bbb46ef8bfa 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-stm32.c
@@ -6,6 +6,7 @@
  * Author:  Alexandre Torgue <alexandre.torgue@st.com> for STMicroelectronics.
  */
 
+#include <linux/bus/stm32_firewall_device.h>
 #include <linux/clk.h>
 #include <linux/kernel.h>
 #include <linux/mfd/syscon.h>
@@ -103,6 +104,7 @@ struct stm32_dwmac {
 	u32 speed;
 	const struct stm32_ops *ops;
 	struct device *dev;
+	struct stm32_firewall firewall;
 };
 
 struct stm32_syscfg_pmcsetr {
@@ -504,11 +506,26 @@ static int stm32_dwmac_probe(struct platform_device *pdev)
 		goto err_remove_config_dt;
 	}
 
+	/* Get stm32 firewall information */
+	ret = stm32_firewall_get_firewall(pdev->dev.of_node, &dwmac->firewall, 1);
+	if (ret)
+		goto err_remove_config_dt;
+
+	if (dwmac->firewall.firewall_id != 0) {
+		ret = stm32_firewall_grant_access_by_id(&dwmac->firewall,
+							dwmac->firewall.firewall_id);
+		if (ret) {
+			dev_err(&pdev->dev, "stm32 firewall grant error:%d\n",
+				ret);
+			goto err_remove_config_dt;
+		}
+	}
+
 	data = of_device_get_match_data(&pdev->dev);
 	if (!data) {
 		dev_err(&pdev->dev, "no of match data provided\n");
 		ret = -EINVAL;
-		goto err_remove_config_dt;
+		goto err_remove_firewall;
 	}
 
 	dwmac->ops = data;
@@ -517,7 +534,7 @@ static int stm32_dwmac_probe(struct platform_device *pdev)
 	ret = stm32_dwmac_parse_data(dwmac, &pdev->dev);
 	if (ret) {
 		dev_err(&pdev->dev, "Unable to parse OF data\n");
-		goto err_remove_config_dt;
+		goto err_remove_firewall;
 	}
 
 	if (stmmac_res.wol_irq && !dwmac->clk_eth_ck) {
@@ -551,6 +568,10 @@ static int stm32_dwmac_probe(struct platform_device *pdev)
 		dev_pm_clear_wake_irq(&pdev->dev);
 		device_set_wakeup_capable(&pdev->dev, false);
 	}
+err_remove_firewall:
+	if (dwmac->firewall.firewall_id != 0)
+		stm32_firewall_release_access_by_id(&dwmac->firewall,
+						    dwmac->firewall.firewall_id);
 err_remove_config_dt:
 	stmmac_remove_config_dt(pdev, plat_dat);
 
@@ -561,6 +582,7 @@ static void stm32_dwmac_remove(struct platform_device *pdev)
 {
 	struct net_device *ndev = platform_get_drvdata(pdev);
 	struct stmmac_priv *priv = netdev_priv(ndev);
+	struct stm32_dwmac *dwmac = priv->plat->bsp_priv;
 
 	stmmac_dvr_remove(&pdev->dev);
 	stm32_dwmac_clk_disable(priv->plat->bsp_priv);
@@ -569,6 +591,10 @@ static void stm32_dwmac_remove(struct platform_device *pdev)
 	device_init_wakeup(&pdev->dev, false);
 
 	phy_power_on(priv->plat->bsp_priv, false);
+
+	if (dwmac->firewall.firewall_id != 0)
+		stm32_firewall_release_access_by_id(&dwmac->firewall,
+						    dwmac->firewall.firewall_id);
 }
 
 static int stm32mp1_suspend(struct stm32_dwmac *dwmac)
-- 
2.50.1

